File: csharp/src/MergeSort.cs

using System.Collections.Generic;

static class MergeSort<T>
{
    public static void Execute(List<T> arr, Func<T, T, bool> comparator)
    {
        if (arr.Count <= 1)
        {
            return;
        }
        Process(arr, 0, arr.Count - 1, comparator);
    }

    private static void Process(List<T> arr, int left, int right, Func<T, T, bool> comparator)
    {
        if (left >= right)
        {
            return;
        }
        int mid = left + ((right - left) / 2);
        Process(arr, left, mid, comparator);
        Process(arr, mid + 1, right, comparator);
        Merge(arr, left, mid, right, comparator);
    }

    private static void Merge(List<T> arr, int left, int mid, int right, Func<T, T, bool> comparator)
    {
        T?[] tmp = new T?[right - left + 1];
        int i = left;
        int j = mid + 1;
        int k = 0;

        while (i <= mid && j <= right)
        {
            if (comparator(arr[i], arr[j]))
            {
                tmp[k++] = arr[i++];
            }
            else
            {
                tmp[k++] = arr[j++];
            }
        }

        for (int idx = i; idx <= mid; ++idx)
        {
            tmp[k++] = arr[idx];
        }

        for (int idx = j; idx <= right; ++idx)
        {
            tmp[k++] = arr[idx];
        }

        for (int idx = 0; idx < tmp.Length; ++idx)
        {
            var nullableItem = tmp[idx];
            if (nullableItem != null)
            {
                arr[left + idx] = nullableItem;
            }
        }
    }
}

File: go/src/sort/mergesort.go
package sort

func MergeSort[T any](arr []T, comparator func(T, T) bool) {
	if len(arr) <= 1 {
		return
	}
	mergeSort(arr, 0, len(arr)-1, comparator)
}

func mergeSort[T any](arr []T, left, right int, comparator func(T, T) bool) {
	if left >= right {
		return
	}
	mid := left + ((right - left) / 2)
	mergeSort(arr, left, mid, comparator)
	mergeSort(arr, mid+1, right, comparator)
	merge(arr, left, mid, right, comparator)
}

func merge[T any](arr []T, left, mid, right int, comparator func(T, T) bool) {
	tmp := make([]T, right-left+1)
	i := left
	j := mid + 1
	k := 0

	for i <= mid && j <= right {
		if comparator(arr[i], arr[j]) {
			tmp[k] = arr[i]
			i++
		} else {
			tmp[k] = arr[j]
			j++
		}
		k++
	}

	for idx := i; idx <= mid; idx++ {
		tmp[k] = arr[idx]
		k++
	}

	for idx := j; idx <= right; idx++ {
		tmp[k] = arr[idx]
		k++
	}

	for idx, val := range tmp {
		arr[left+idx] = val
	}
}
